### PR TITLE
FOUNDATIONS: Modify Intern

### LINKED ISSUE #50: FOUNDATIONS: Modify Intern

### PR DESCRIPTION
**Pull request by Pwongchaiya**
Closes: #50

### CODE REVIEW

**Comment by glhays** [comment]
@Pwongchaiya please fix the issue closing in your comment above. The format has to be as follows.
Closes: #issuenumber

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.Modify.cs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+namespace InternTrack.Core.Api.Tests.Unit.Services.Foundations.Interns
+{
+    public partial class InternServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowCriticalDependencyExceptionOnModifyIfSqlErrorOccursAndLogItAsync()
+        {
+            // given
+            Intern randomIntern = CreateRandomIntern();
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.Modify.cs:24]: call this someIntern. We don't care about randomness in this unit test

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.Modify.cs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnModifyIfDatabaseUpdateExceptionOccursAndLogItAsync()
+        {
+            // given
+            Intern randomIntern = CreateRandomIntern();
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.Modify.cs:71]: random -> some

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.Modify.cs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                broker.SelectInternByIdAsync(randomIntern.Id),
+                    Times.Never);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnModifyIfDbUpdateConcurrencyExceptionOccursAndLogItAsync()
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.Modify.cs:115]: dbUpdateConcurrency exception should come before dbUpdate exception

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Logic.Modify.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public partial class InternServiceTests
+    {
+        [Fact]
+        public async Task ShouldModifyInternAsync()
+        {
+            // given
+            DateTimeOffset randomDate = GetRandomDateTime();
+            Intern randomIntern = CreateRandomIntern();
+            Intern inputIntern = randomIntern;
+            Intern StorageIntern = inputIntern.DeepClone();
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Logic.Modify.cs:25]: Lower case Storage

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Logic.RetrieveById.cs [hunk 5]
```diff
@@ -25,7 +25,7 @@ public async Task ShouldRetrieveInternByIdAsync()
             Intern storageIntern = randomIntern;
             Intern expectedIntern = storageIntern.DeepClone();
 
-            this.storageBrokerMock.Setup(broker => 
+            this.storageBrokerMock.Setup(broker =>
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Logic.RetrieveById.cs:28]: Try not to cleanup files that are not related to Modify feature

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.Modify.cs [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+                broker.SelectInternByIdAsync(It.IsAny<Guid>()),
+                    Times.Never);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.UpdateInternAsync(It.IsAny<Intern>()),
+                    Times.Never);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.Modify.cs:56]: Order of these are incorrect

**Review by ElbekDeveloper — CHANGES_REQUESTED** [review]

**Comment by glhays** [comment]
@Pwongchaiya you need the colon after closes.
You also haven't git verified it appears.

File: InternTrack.Core.Api/Models/Interns/Exceptions/InternNotFoundException.cs
Comment (glhays) [inline InternTrack.Core.Api/Models/Interns/Exceptions/InternNotFoundException.cs:1]: This InternNotFoundException is duplicated with NotFoundInternException. Consider removing.
